src/main/scala/variant/Horde.scala

```diff
@@ -48,19 +48,61 @@ case object Horde extends Variant(
+    val hordePos = board.occupation(Color.white) // may include promoted pieces
+    val mateInOne = hordePos.size == 1 && hordePos.forall(pos => pieceThreatened(board, Color.black, pos, (_ => true)))
+    !mateInOne && notKingBoard.actors.values.forall(actor => actor.moves.isEmpty)
+  }
+
+  /**
+   * In horde chess, black can win unless a fortress stalemate is unavoidable.
+   *  Auto-drawing the game should almost never happen, but it did in https://lichess.org/xQ2RsU8N#121
+   */
+  override def insufficientWinningMaterial(board: Board) = hordeClosedPosition(board)
 
   /**
-   * In horde chess, white cannot win on * V K or [BN]{2} v K or just one piece since they don't have a king
-   * for support.
+   * In horde chess, the horde cannot win on * V K or [BN]{2} v K or just one piece
+   * since they lack a king for checkmate support.
+   * Technically there are some positions where stalemate is unavoidable which
+   * this method does not detect; however, such are trivial to premove.
    */
-  override def insufficientWinningMaterial(board: Board, color: Color) = {
-    color == Color.white && board.piecesOf(Color.white).size == 1 ||
-      board.piecesOf(Color.white).size == 2 &&
-      board.piecesOf(Color.white).forall(_._2.isMinor)
+  override def insufficientWinningMaterial(board: Board, color: Color): Boolean = {
+    if (color == Color.white) {
+      lazy val notKingPieces = InsufficientMatingMaterial.nonKingPieces(board)
+      val horde = board.piecesOf(Color.white)
+      lazy val hordeBishopSquareColors = horde.filter(_._2.is(Bishop)).toList.map(_._1.color).distinct
+      lazy val hordeRoles = horde.map(_._2.role)
+      lazy val army = board.piecesOf(Color.black)
+      lazy val armyRooks = army.filter(p => p._2.is(Pawn) || p._2.is(Rook))
+      lazy val armyBishops = army.filter(p => p._2.is(Pawn) || p._2.is(Bishop))
+      lazy val armyKnights = army.filter(p => p._2.is(Pawn) || p._2.is(Knight))
+      lazy val armyPawns = army.filter(_._2.is(Pawn))
+      lazy val armyNonQueens = army.filter(_._2.isNot(Queen))
+      lazy val armyNonRooks = army.filter(p => p._2.isNot(Queen) && p._2.isNot(Rook))
+      lazy val armyNonBishops = army.filter(p => p._2.isNot(Queen) && p._2.isNot(Bishop))
+      lazy val armyBishopSquareColors = armyBishops.toList.map(_._1.color).distinct
+      if (horde.size == 1) {
+        if (hordeRoles == List(Knight)) {
+          if (army.size < 4 || armyNonRooks.isEmpty || armyNonBishops.isEmpty || armyPawns.isEmpty && (armyNonBishops.size + armyBishopSquareColors.size < 4)) return true
+        } else if (hordeRoles == List(Bishop)) {
+          if (!notKingPieces.exists(_._2.role != Bishop) && notKingPieces.map(_._1.color).distinct.size == 1) return true
+        } else if (hordeRoles == List(Rook)) {
+          if (army.size < 3 || armyRooks.isEmpty || armyKnights.isEmpty) return true
+        } else {
+          if (armyRooks.isEmpty) return true
```

Comment:
scalachess doesn't use `return`, which is considered a code smell in scala. You may want to use that instead:

```
hordeRoles match {
  case List(Knight) => ...
  case List(Bishop) => ...
  case _ => ...
}
```